static/js/components/tiles/map_tile.tsx

```diff
@@ -554,108 +574,151 @@ export function draw(
-  for (const placeData of chartData.placeData) {
+
+  for (const layer of chartData.layerData) {
+    // Build mapping of all values and metadata for a place, to show in tooltip
+    for (const place in layer.dataValues) {
+      if (!Object.keys(allDataValues).includes(place)) {
+        allDataValues[place] = {};
+      }
+      allDataValues[place][layer.variable.statVar] = layer.dataValues[place];
+
+      if (!Object.keys(allMetadataValues).includes(place)) {
+        allMetadataValues[place] = {};
+      }
+      allMetadataValues[place][layer.variable.statVar] = layer.metadata[place];
+    }
+    // Build variable -> values mapping for color scale calculations
+    // Assumes each place and variable combination will only ever have 1 unique
+    // value.
+    allValuesByVariable[layer.variable.statVar] = Object.values(
+      layer.dataValues
+    );
+    variableNames[layer.variable.statVar] = layer.variable.name;
+    units[layer.variable.statVar] = layer.unit;
+
     // Use border data to calculate projection if using borders.
     // This prevents borders from being cutoff when enclosed places don't
     // provide wall to wall coverage.
     const shouldUseBorderData =
-      placeData.enclosedPlaceType &&
-      shouldShowBorder(placeData.enclosedPlaceType) &&
-      !_.isEmpty(placeData.borderGeoJson);
+      layer.enclosedPlaceType &&
+      shouldShowBorder(layer.enclosedPlaceType) &&
+      !_.isEmpty(layer.borderGeoJson);
     projectionGeoJsons.push(
-      shouldUseBorderData ? placeData.borderGeoJson : placeData.geoJson
+      shouldUseBorderData ? layer.borderGeoJson : layer.geoJson
     );
     if (shouldUseBorderData) {
-      bordersToDraw.push(placeData.borderGeoJson);
+      bordersToDraw.push(layer.borderGeoJson);
     }
   }
 
+  // Calculate color scales for each variable and add legends
+  let totalLegendWidth = 0;
+  const legendData: {
+    colorScale: d3.ScaleLinear<number, number, never>;
+    unit: string;
+    label?: string;
+  }[] = [];
+  for (const variable in allValuesByVariable) {
+    // calculate color scale based on max/min values across all layers
+    const dataValues = allValuesByVariable[variable];
+    const colorScale = getColorScale(
+      variable,
+      d3.min(dataValues),
+      d3.mean(dataValues),
+      d3.max(dataValues),
+      undefined,
+      undefined,
+      props.colors
+    );
+    // add colorScale to layers
+    for (const layer of chartData.layerData) {
+      if (layer.variable.statVar === variable) {
+        layer.colorScale = colorScale;
+      }
+    }
+    legendData.push({
+      colorScale,
+      unit: units[variable],
+      label: variableNames[variable],
+    });
+  }
+
+  // add legend
+  totalLegendWidth += generateLegendSvg(
+    legendContainer,
+    props.svgChartHeight,
+    legendData,
+    10
+  );
+  const chartWidth = (svgWidth || svgContainer.offsetWidth) - totalLegendWidth;
+  const getTooltipHtml = (place: NamedPlace) => {
```

Comment:
make a helper function for this? & can move all the mappings for tooltips there. this function is very large ..

Reply:
You're right the function is too large now. Moved the tooltip logic to a helper function in the utils file, as well as the projection and color scale calculations.